Tests/FluentAssertions.Specs/OccurrenceConstraintSpecs.cs

```diff
@@ -0,0 +1,383 @@
+﻿using System;
+using FluentAssertions.Execution;
+using Xunit;
+using Xunit.Sdk;
+
+namespace FluentAssertions.Specs
+{
+    public class OccurrenceConstraintSpecs
+    {
+        #region At Least
+
+        [InlineData(0, true)]
+        [InlineData(1, false)]
+        [Theory]
+        public void At_least_once_should_be_bigger_than_zero(int actual, bool shouldThrow)
+        {
+            // Arrange
+            OccurrenceConstraint constraint = AtLeast.Once();
+
+            // Act
+            Action act = () => Execute.Assertion
+                .ForConstraint(constraint, actual)
+                .FailWith($"Expected occurrence to be {constraint.Mode} {constraint.ExpectedCount}, but it was {actual}");
+
+            // Assert
+            if (shouldThrow)
```

Comment:
I don't think we should be compressing tests by introducing conditional logic.
It's only an additional five lines to split the test into a positive and negative tests.

```cs
[InlineData(1)]
[InlineData(2)]
[Theory]
public void At_least_once_is_bigger_than_zero(int occurrences)
{
    // Arrange
    OccurrenceConstraint constraint = AtLeast.Once();

    // Act / Assert
    Execute.Assertion
        .ForConstraint(constraint, occurrences)
        .FailWith("");
}

[Fact]
public void At_least_once_is_not_bigger_than_zero()
{
    // Arrange
    OccurrenceConstraint constraint = AtLeast.Once();

    // Act
    Action act = () => Execute.Assertion
        .ForConstraint(constraint, 0)
        .FailWith($"Expected occurrence to be {constraint.Mode} {constraint.ExpectedCount}, but it was 0");

    // Assert
    act.Should().Throw<XunitException>()
            .WithMessage("Expected occurrence to be at least 1, but it was 0*");
}
```

If we _really_ wanted to compress tests we could cover all scenarios with this (modulo the relaxed failure message assertions)

```cs
public static object[][] PassingConstraints() => new object[][]
{
    new object[] { AtLeast.Once(), 1 },
    new object[] { AtLeast.Once(), 2 },
    new object[] { AtLeast.Twice(), 2 },
    new object[] { AtLeast.Twice(), 3 },
    new object[] { AtLeast.Thrice(), 3 },
    new object[] { AtLeast.Thrice(), 4 },
    new object[] { AtLeast.Times(4), 4 },
    new object[] { AtLeast.Times(4), 5 },

    new object[] { AtMost.Once(), 0 },
    new object[] { AtMost.Once(), 1 },
    new object[] { AtMost.Twice(), 1 },
    new object[] { AtMost.Twice(), 2 },
    new object[] { AtMost.Thrice(), 2 },
    new object[] { AtMost.Thrice(), 3 },
    new object[] { AtMost.Times(4), 3 },
    new object[] { AtMost.Times(4), 4 },

    new object[] { Exactly.Once(), 1 },
    new object[] { Exactly.Twice(), 2 },
    new object[] { Exactly.Thrice(), 3 },
    new object[] { Exactly.Times(4), 4 },

    new object[] { LessThan.Twice(), 1 },
    new object[] { LessThan.Thrice(), 2 },
    new object[] { LessThan.Times(4), 3 },

    new object[] { MoreThan.Once(), 2 },
    new object[] { MoreThan.Twice(), 3 },
    new object[] { MoreThan.Thrice(), 4 },
    new object[] { MoreThan.Times(4), 5 },
};

[Theory]
[MemberData(nameof(PassingConstraints))]
public void Occurrence_constraint_passes(OccurrenceConstraint constraint, int occurrences)
{
    // Act / Assert
    Execute.Assertion
        .ForConstraint(constraint, occurrences)
        .FailWith("");
}

public static object[][] FailingConstraints() => new object[][]
{
    new object[] { AtLeast.Once(), 0 },
    new object[] { AtLeast.Twice(), 1 },
    new object[] { AtLeast.Thrice(), 2 },
    new object[] { AtLeast.Times(4), 3 },

    new object[] { AtMost.Once(), 2 },
    new object[] { AtMost.Twice(), 3 },
    new object[] { AtMost.Thrice(), 4 },
    new object[] { AtMost.Times(4), 5 },

    new object[] { Exactly.Once(), 0 },
    new object[] { Exactly.Once(), 2 },
    new object[] { Exactly.Twice(), 1 },
    new object[] { Exactly.Twice(), 3 },
    new object[] { Exactly.Thrice(), 2 },
    new object[] { Exactly.Thrice(), 4 },
    new object[] { Exactly.Times(4), 3 },
    new object[] { Exactly.Times(4), 5 },

    new object[] { LessThan.Twice(), 2 },
    new object[] { LessThan.Twice(), 3 },
    new object[] { LessThan.Thrice(), 3 },
    new object[] { LessThan.Thrice(), 4 },
    new object[] { LessThan.Times(4), 4 },
    new object[] { LessThan.Times(4), 5 },

    new object[] { MoreThan.Once(), 0 },
    new object[] { MoreThan.Once(), 1 },
    new object[] { MoreThan.Twice(), 1 },
    new object[] { MoreThan.Twice(), 2 },
    new object[] { MoreThan.Thrice(), 2 },
    new object[] { MoreThan.Thrice(), 3 },
    new object[] { MoreThan.Times(4), 3 },
    new object[] { MoreThan.Times(4), 4 },
};

[Theory]
[MemberData(nameof(FailingConstraints))]
public void Occurrence_constraint_fails(OccurrenceConstraint constraint, int occurrences)
{
    // Act
    Action act = () => Execute.Assertion
        .ForConstraint(constraint, occurrences)
        .FailWith($"Expected occurrence to be {constraint.Mode} {constraint.ExpectedCount}, but it was {occurrences}");

    // Assert
    act.Should().Throw<XunitException>()
            .WithMessage("Expected occurrence to be *, but it was *");
}
```

Reply:
I really like the `MemberData` example. If @dennisdoomen approves, I will modify it into this; otherwise I will split all the tests.

Reply:
I think this is one of those cases where it helps.